lib/vega_lite/data.ex

```diff
@@ -203,8 +258,71 @@ defmodule VegaLite.Data do
     opts |> Keyword.put_new(:type, :quantitative) |> Keyword.put_new(:aggregate, :count)
   end
 
-  defp density_heatmap_defaults(_field, opts) do
-    opts
+  defp density_heatmap_defaults(_field, opts), do: opts
+
+  ### Specialized - builders
+
+  defp build_heatmap_layers(vl, data, {cols, fields, used_fields}) do
+    text_fields = Keyword.take(fields, [:text, :x, :y])
+    rect_fields = Keyword.delete(fields, :text)
+
+    text_layer = if fields[:text], do: [encode_layer(cols, :text, text_fields)], else: []
+    rect_layer = [encode_layer(cols, :rect, rect_fields)]
+
+    vl
+    |> Vl.data_from_values(data, only: used_fields)
+    |> Vl.layers(rect_layer ++ text_layer)
+  end
+
+  defp build_jointplot(vl, data, used_fields, main_chart, {x_hist, y_hist}) do
+    root_visuals = %{"bounds" => "flush", "spacing" => 15}
+
+    vl
+    |> Map.update!(:spec, &Map.merge(&1, root_visuals))
+    |> Vl.data_from_values(data, only: used_fields)
+    |> Vl.concat(
+      [x_hist, Vl.new(spacing: 15, bounds: :flush) |> Vl.concat([main_chart, y_hist])],
+      :vertical
+    )
+  end
+
+  defp build_main_jointplot(data, :density_heatmap, fields, opts) do
+    Vl.new(opts)
+    |> density_heatmap(data, fields)
+    |> Map.update!(:spec, &Map.delete(&1, "data"))
```

Comment:
Unfortunately, if we discard the data, it means we will do duplicate processing because `data_from_values` traverses `data` to create a data-structure serializable to JSON. I think that, instead of discard the data, you should pass `extra_fields: ...` to `density_heatmap` and `chart` to make sure the data is converted and correct at once. WDYT?